syntax = "proto3";

option java_multiple_files = true;
option java_package = "edu.pucmm.grpc";
option java_outer_classname = "UrlServiceProto";

// Mensaje de registro de accesos
message AccessLog {
  int64 timestamp = 1;
  string ip = 2;
  string user = 3;
  string platform = 4;
  string browser = 5;
  string clientDomain = 6;
}

// Mensaje que representa la URL con sus estadísticas
message UrlRecord {
  string fullUrl = 1;
  string shortUrl = 2;
  int64 creationDate = 3;
  int32 accessCount = 4;
  repeated AccessLog accessLogs = 5;
  string previewImageBase64 = 6;
}

// Para la operación de listar URLs
message ListUserUrlsRequest {
  string username = 1;
}
message ListUserUrlsResponse {
  repeated UrlRecord urls = 1;
}

// Para la operación de crear registro
message CreateUrlRecordRequest {
  string username = 1;
  string originalUrl = 2;
}
message CreateUrlRecordResponse {
  UrlRecord urlRecord = 1;
}

// Definición del servicio gRPC
service UrlExtendedService {
  rpc ListUserUrls(ListUserUrlsRequest) returns (ListUserUrlsResponse);
  rpc CreateUrlRecord(CreateUrlRecordRequest) returns (CreateUrlRecordResponse);
}
